src/test/java/org/jenkinsci/plugins/gitclient/GitAPITestCase.java

```diff
@@ -405,17 +400,49 @@ private ObjectId getMirrorHead() throws IOException, InterruptedException
     protected String localMirror() throws IOException, InterruptedException {
         File base = new File(".").getAbsoluteFile();
         for (File f=base; f!=null; f=f.getParentFile()) {
-            if (new File(f,"target").exists()) {
-                File clone = new File(f, "target/clone.git");
-                if (!clone.exists()) {  // TODO: perhaps some kind of quick timestamp-based up-to-date check?
-                    w.launchCommand("git", "clone", "--mirror", "https://github.com/jenkinsci/git-client-plugin.git", clone.getAbsolutePath());
+            File targetDir = new File(f, "target");
+            if (targetDir.exists()) {
+                String cloneDirName = "clone.git";
+                File clone = new File(targetDir, cloneDirName);
+                if (!clone.exists()) {
+                    Path tempClonePath = Files.createTempDirectory(targetDir.toPath(), "clone-");
+                    w.launchCommand("git", "clone", "--reference", f.getCanonicalPath(), "--mirror", "https://github.com/jenkinsci/git-client-plugin", tempClonePath.toFile().getAbsolutePath());
+                    if (!clone.exists()) { // Still a race condition, but a narrow race handled by Files.move()
+                        renameAndDeleteDir(tempClonePath, cloneDirName);
+                    } else {
+                        Util.deleteRecursive(tempClonePath.toFile());
```

Comment:
When maven runs tests, it starts them in parallel.  Each of the tests which use WorkspaceWithRepo will attempt to create the local cache in `target/clone.git`.  One of those "wins the race" and moves the separate clone to the name `target/clone.git`.  The others "lose the race" and either fail their attempt to move or take this branch where they detect the existence of `target/clone.git` and delete their clone rather than attempting to move it.

Reply:
I'll add a comment over that line to explain why it exists.  This was the line that I most frequently hit while researching the test instability on a machine with many cores.

Reply:
Comment was added in the implementation that is part of WorkspaceWithRepo, since that implementation will continue to exist even after the JUnit 3 tests have all been moved to other JUnit 4 and other test classes.

Reply:
This makes it pretty clear, thanks for adding the comment.